File: 303.range-sum-query-immutable.py
class NumArray:

    def __init__(self, nums: List[int]):
        self.nums_sum = [0]*len(nums)
        for i, _ in enumerate(nums):
            self.nums_sum[i] = self.nums_sum[i-1] + (nums[i] if i > 0 else nums[i])

    def sumRange(self, left: int, right: int) -> int:
        return self.nums_sum[right] - self.nums_sum[left] + self.nums_sum[left] - (self.nums_sum[left-1] if left > 0 else 0)

File: 303.range-sum-query-immutable.c

typedef struct {
	int* nums;
	} NumArray;

NumArray* numArrayCreate(int* nums, int numsSize) {
	int i;
	NumArray* na = malloc(sizeof(NumArray));
	na->nums = malloc(sizeof(int) * (numsSize + 1));

	int sum = 0;
	na->nums[0] = sum;
	for (i = 1; i <= numsSize; ++i) {
		sum += nums[i - 1];
		na->nums[i] = sum;
		}
	return na;
	}

int numArraySumRange(NumArray* obj, int left, int right) {
	return obj->nums[right + 1] - obj->nums[left];
	}

void numArrayFree(NumArray* obj) {
	free(obj->nums);
	free(obj);
	}
